keps/sig-cli/0031-datadrivencommands.md

```diff
@@ -0,0 +1,412 @@
+	//          matchLabels:
+	//            app: {{index .Flags.Strings "name"}}
+	//        template:
+	//          metadata:
+	//            labels:
+	//              app: {{index .Flags.Strings "name"}}
+	//          spec:
+	//            containers:
+	//            - name: {{index .Flags.Strings "name"}}
+	//              image: {{index .Flags.Strings "image"}}
+	//
+	// +optional
+	BodyTemplate string `json:"bodyTemplate,omitempty"`
+
+	// SaveResponseValues are values read from the response and saved in {{index .Responses.Strings "flag-name"}}.
+	// They may be used in the ResourceCommand.Output go-template.
+	//
+	// Example:
+	//		- name: responsename
+	//        jsonPath: "{.metadata.name}"
+	//
+	// +optional
+	SaveResponseValues []ResponseValue `json:"saveResponseValues,omitempty"`
+}
+
+// Flag defines a cli flag that should be registered and available in request / output templates
+type Flag struct {
+	Type FlagType `json:"type"`
+
+	Name string `json:"name"`
+
+	Description string `json:"description"`
+
+	// +optional
+	StringValue string `json:"stringValue,omitempty"`
+
+	// +optional
+	StringSliceValue []string `json:"stringSliceValue,omitempty"`
+
+	// +optional
+	BoolValue bool `json:"boolValue,omitempty"`
+
+	// +optional
+	IntValue int32 `json:"intValue,omitempty"`
+
+	// +optional
+	FloatValue float64 `json:"floatValue,omitempty"`
+}
+
+// ResponseValue defines a value that should be parsed from a response and available in output templates
+type ResponseValue struct {
+	Name     string `json:"name"`
+	JsonPath string `json:"jsonPath"`
+}
+
+type FlagType string
+
+const (
+	STRING       FlagType = "String"
+	BOOL                  = "Bool"
+	FLOAT                 = "Float"
+	INT                   = "Int"
+	STRING_SLICE          = "StringSlice"
+)
+
+type Command struct {
+	// Use is the one-line usage message.
+	Use string `json:"use"`
+
+	// Path is the path to the sub-command.  Omit if the command is directly under the root command.
+	// +optional
+	Path []string `json:"path,omitempty"`
+
+	// Short is the short description shown in the 'help' output.
+	// +optional
+	Short string `json:"short,omitempty"`
+
+	// Long is the long message shown in the 'help <this-command>' output.
+	// +optional
+	Long string `json:"long,omitempty"`
+
+	// Example is examples of how to use the command.
+	// +optional
+	Example string `json:"example,omitempty"`
+
+	// Deprecated defines, if this command is deprecated and should print this string when used.
+	// +optional
+	Deprecated string `json:"deprecated,omitempty"`
+
+	// Flags are the command line flags.
+	//
+	// Example:
+	// 		  - name: namespace
+	//    		type: String
+	//    		stringValue: "default"
+	//    		description: "deployment namespace"
+	//
+	// +optional
+	Flags []Flag `json:"flags,omitempty"`
+
+	// SuggestFor is an array of command names for which this command will be suggested -
+	// similar to aliases but only suggests.
+	SuggestFor []string `json:"suggestFor,omitempty"`
```

Comment:
While I know why this looks like the cobra struct, something about bothers me.

Reply:
We can iterate on the precise API format and capabilities.

Reply:
One of the main goals of the PoC was to demonstrate that Cobra commands could be represented as declarative data published by the cluster.  Having proven that, re-thinking the representation and capabilities sounds like a good plan.